### ISSUE TITLE
Rollup: Allow definition of jQuery style

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#632]
In order to make sure we have the correct style rules, I'd like to use the jQuery style guide as a reference. If we can implement the jQuery style guide in ESLint, then we will have a good set of style rules as a basis.

- [x] Identation check (#1022)
- [x] No trailing whitespace
- [x] Unary operators must not have spaces (`space-unary-ops`)
- [x] Ternary statements must have spaces before and after both `?` and `:` (`space-infix-ops`)
- [x] Maximum line length (`max-len`)
- [x] Require braces for control statements (`curly`)
- [x] Require multiple lines for control statements (`curly`)
- [x] Commas cannot have a preceding space (#628)
- [x] Semicolons cannot have a preceding space (`no-space-before-semi`)
- [x] Semicolons only at end of line (`semi`)
- [x] The `:` in property names should have no preceding space (#1280)
- [x] No filler spaces in empty constructs like `[]`, `{}`
- [x] New line at end of file
- [ ] If entire file is wrapped in closure, the body should not be indented
- [x] Spaces inside of parens (#627)
- [x] Spaces inside of square brackets (#631)
- [x] Line breaks must occur after an operator (#1405)
- [x] Always use semicolons (`semi`)
- [x] Use camel case (`camelcase`)
- [x] Allow non-capital constructor names
- [x] Align `case` with `switch`

**Comment by gkz** [comment]
Issues #1022 (indentation check) and #1405 (line breaks after operator) are closed - only a couple of issues left!

**Comment by gyandeeps** [comment]
@nzakas case with switch is done with indent rule rewrite.
I am not sure if we want to do this one: `If entire file is wrapped in closure, the body should not be indented`

**Comment by nzakas** [comment]
For completeness, I think we should.

**Comment by gyandeeps** [comment]
Let do the last pending item, Should we add an option in indent rule? if yes then lets come up with a name.

CC @nzakas @BYK

**Comment by nzakas** [comment]
How does JSCS do this?

**Comment by gyandeeps** [comment]
They dont have an option for it. Quick code scan tells me that they do it by default, I guess.
https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/validate-indentation.js#L387-L388

**Comment by nzakas** [comment]
Looks like it. I don't think we want to do it by default. We need an option for it, though.

**Comment by ndaidong** [comment]
How about "no-invalid-this". While using jQuery, it's normal to write:

  $('#somebutton').click(function(){
    var pid = $(this).attr('pid');
    doAction(pid);
  });

With "no-invalid-this" enabled, that will become an error.

**Comment by mysticatea** [comment]
There is an workaround.

```js
$('#somebutton').click(/* @this HTMLElement */ function(){
    var pid = $(this).attr('pid');
    doAction(pid);
});
```

**Comment by ndaidong** [comment]
nicely, thank you @mysticatea :+1:

**Comment by BYK** [comment]
@nzakas @gyandeeps - I feel like `indent` requires a rewrite so not sure if adding a new option for this is the best thing at the time.

**Comment by nzakas** [comment]
Unless the rewrite is coming soon, then adding an option is the best path forward. I'd like to finally close this issue.

**Comment by babbelnedd** [comment]
@mysticatea how about 

    self.emit('foo', onFoo($(this)));

**Comment by mjschlot** [comment]
I realize that the jQuery question is a few months old, but I wanted to post some details for anybody who finds this page using a Google search. Instead of using '/* @this HTMLElement */' to avoid linting errors there is an alternative.

With event listeners like 'click' you can use 'event.currentTarget'.
http://api.jquery.com/event.currentTarget/

With jQuery functions like 'each' there are is a second argument that can be used in place of 'this':
http://api.jquery.com/each/

**Comment by nzakas** [comment]
I'm going to close this issue. Since we only had one thing missing from the list, and no one seems to know if it's important or not, I'd rather move on and see if anyone files an issue for it.

**Comment by alberto** [comment]
It's now covered by `outerIIFEBody` in `indent`. I'm editing the original comment to reflect that.
